interpreter/evalstate.go

```diff
@@ -31,17 +33,21 @@ type MutableEvalState interface {
 }
 
 // NewEvalState returns a MutableEvalState.
-func NewEvalState() MutableEvalState {
-	return &defaultEvalState{make(map[int64]ref.Value)}
+func NewEvalState(instructionCount int64) *defaultEvalState {
+	return &defaultEvalState{exprCount: instructionCount,
+		exprValues: make([]ref.Value, instructionCount, instructionCount)}
 }
 
 type defaultEvalState struct {
-	exprValues map[int64]ref.Value
+	exprCount  int64
+	exprValues []ref.Value
 }
 
 func (s *defaultEvalState) Value(exprId int64) (ref.Value, bool) {
-	object, found := s.exprValues[exprId]
-	return object, found
+	if exprId >= 0 && exprId < s.exprCount {
+		return s.exprValues[exprId], true
+	}
+	return nil, false
```

Comment:
why is this not treated like an error?

Reply:
Does this check imply/impose that the expressions id have to be densely assigned? It looks like a better way could be to check the existence of exprid in the map.